### ISSUE TITLE
Local variable interpolation error is silent

### ISSUE THREAD

**Issue by Chili-Man** [hashicorp/terraform#16682]
Terraform should have failed on the bad interpolation of a local variable, but instead failed elsewhere.

### Terraform Version
```
Terraform v0.11.0
+ provider.aws v1.3.0
+ provider.template v1.0.0
```
### Terraform Configuration Files
```hcl
provider "aws" {
  region  = "us-west-2"
}

data "aws_availability_zones" "datacenters" {}

locals {
  azs       = ["${aws_availability_zones.datacenters.names}"]  # error is here, its missing the `data.`
  azs_count = "${length(local.azs)}"
}

data "template_file" "public" {
  count = "${local.azs_count}"
  template = "$${cidrsubnet("10.0.0.0/8", 8, count.index)}"
}
```

### Debug Output
https://gist.github.com/Chili-Man/14344ea5f3e0ea161724b4024fd5f9ce

### Expected Behavior
Terraform should have failed with an error about the bad interpolation on the local variable azs

### Actual Behavior
I got the following error instead:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_availability_zones.datacenters: Refreshing state...

Error: Error refreshing state: 1 error(s) occurred:

* data.template_file.public: data.template_file.public: value of 'count' cannot be computed
```

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform plan`

**Comment by apparentlymart** [comment]
Hi @Chili-Man! Thanks for reporting this, and sorry for the incorrect behavior.

Is this a new error for 0.11.0, or have you seen the same behavior on a 0.10 release before? I ask because I don't think we significantly changed the handling of local values in 0.11, so knowing if it's a 0.11 regression would probably help narrow down the cause.

(If you don't have a 0.10 release handy to test then no worries... just wondering if you'd already tried it.)

**Comment by Chili-Man** [comment]
hey @apparentlymart , I just tested it on 0.10.8 and the same error happens, so it doesn't seem to be a 0.11.0 regression.

**Comment by apparentlymart** [comment]
Hi @Chili-Man! Sorry for the long silence here.

I'm pleased to report that this is now fixed on master, as verified by running your configuration through the v0.12.0-alpha1 release:

```
$ terraform apply

Error: Reference to undeclared resource

  on error-in-local.tf line 8, in locals:
   8:   azs = ["${aws_availability_zones.datacenters.names}"] # error is here, its missing the `data.`

A resource "aws_availability_zones" "datacenters" has not been declared in the
root module
```

This fix will also be included in the forthcoming v0.12.0 final release. Thanks again for reporting this, and thanks for your patience while we worked through this.
